implementations/rust/ockam/ockam_core/src/api.rs

```diff
@@ -5,15 +5,62 @@ use crate::compat::rand;
 use crate::compat::vec::Vec;
 use crate::errcode::{Kind, Origin};
 use crate::Result;
+use alloc::collections::btree_map::Entry;
 use core::fmt::{self, Display, Formatter};
 use minicbor::encode::{self, Encoder, Write};
 use minicbor::{Decode, Decoder, Encode};
+use once_cell::race::OnceBox;
 use tinyvec::ArrayVec;
 
 #[cfg(feature = "tag")]
 use crate::TypeTag;
 
-pub const SCHEMA: &str = core::include_str!("schema.cddl");
+use cddl_cat::{context::BasicContext, flatten::flatten, parse_cddl};
+
+const SCHEMA: &str = core::include_str!("schema.cddl");
+
+fn schema(cddl_schema: &str) -> Result<BasicContext> {
+    let cddl = parse_cddl(cddl_schema).map_err(|e| crate::Error::new(Origin::Core, Kind::Io, e))?;
+    let cddl = flatten(&cddl).map_err(|e| crate::Error::new(Origin::Core, Kind::Io, e))?;
+    Ok(BasicContext::new(cddl))
+}
+
+pub fn merged_cddl(cddl_schemas: &[&str]) -> Result<&'static BasicContext> {
+    static INSTANCE: OnceBox<BasicContext> = OnceBox::new();
```

Comment:
This is not exactly how you would use this.

Note from the [example](https://github.com/build-trust/ockam/blob/7e62aed2222aaef50c5c3bddb589187d626be0de/implementations/rust/ockam/ockam_abac/src/parser.rs#L13) I posted on the previous command that the once_cell is used in isolation (behind a function to make it easy to access it).

Since each crate has their schemas and are known at compile time, the idea is that we can create a static reference to the compiled cddl *of each crate*.

If each crate has their own static cddl, you could then rewrite this:

```rust
let cddl_context = merged_cddl(&[])?;
assert_request_match(schema, &buf, cddl_context);
```

to:

```rust
assert_request_match(schema, &buf, cddl());
```

The static cdd, on the `ockam_api` crate for example, can be built as follows:

```rust
fn cddl() -> &'static BasicContext {
    static INSTANCE: OnceBox<BasicContext> = OnceBox::new();
    INSTANCE.get_or_init(|| {
        Box::new(merged_cddl(&[])).unwrap())
    })
}
```

Reply:
Thanks for the suggestions. But how would you handle the errors that `merged_cddl` might produce? Because `assert_request_match` only returns (). Should they be outputted via tracing::error! ?

Also can you maybe explain how your suggestion would differ from mine? Because in the end in my approach we would also just get the `&'static BasicContext` assigned to `cddl_context` and then passed to `assert_request_match` in 
```rust
let cddl_context = merged_cddl(&[])?;
assert_request_match(schema, &buf, cddl_context);
```

How does this differ from calling a function that wraps the `Result` output from `merged_cddl` in the `OnceBox` and calling and passing that to `assert_request_match`:
```rust
fn cddl() -> &'static BasicContext {
    static INSTANCE: OnceBox<BasicContext> = OnceBox::new();
    INSTANCE.get_or_init(|| {
        Box::new(merged_cddl(&[])).unwrap())
    })
}
``` 
Also wouldn't the above function return a `&'static Result<BasicContext>` due to merged_cddl"s return value?

If i call my version of merged_cddl (with OnceBox) in ockam_api for example this will create it's own cddl at compile time, just as it would for ockam_core or any other crate that imports `merged_cddl`, no? 

Sorry for further delay on this. I think this is the last question that I had and hopefully we can merge this soon. Thanks for all the help thus far @adrianbenavides

Reply:
> But how would you handle the errors that merged_cddl might produce?

We can't handle that error, since the cddl's are compiled as part of the binary. If there is an error with the cddl's we better fail fast, so it's okay to unwrap that error.

> Also can you maybe explain how your suggestion would differ from mine?

There are two differences:

1. Instead of having to merge cddls **everytime** before doing an assert_match with `merged_cddl(&[])?` you can just load the **pre-merged** cddl with `cddl()`. 
2. The second benefit is that you don't have to explicitly write the cddl's you want to merge (the `merged_cddl` argument), since that's constant within a crate.

> Or is the output of both approaches kind of the same, but you would suggest that we use the approach that you suggested?

I think the output should be the same, but my suggestion has the previous benefits.

Reply:
hi @adrianbenavides thanks. 
That of course makes sense. Not sure why I didn't catch that myself. 
In your solution the check for an existing value in the `OnceBox` happens before the call to `merged_cddl` (and calls `merged_cddl` on the first time), while in my version `merged_cddl` get's called everytime and afterwards `OnceBox` `get_or_init` get's called, leading to unnecessary merging of the cddl everytime. 

I changed it to the solution that you suggested. I am not a big fan of having to define the `cddl` function in every crate now, but i guess that's not too bad considering the advantages of `OnceBox`. 

One thing that I just caught is that the `schema.cddl` file that I split, I accidentally added it to `ockam_node` instead of `ockam_api`. `ockam_node` only needs the types defined in the schema from `ockam_core` I assume. But `ockam_api` needs other ones too so I moved the file to `ockam_api`. I am not sure where I should import the `SCHEMA` and create the `cddl()` function in `ockam_api` so I did in `ockam_api::authenticator::direct` because that's where it's used. 

Tests in `ockam_api::cloud::project::tests::schema::create_project` and other schema tests in `ockam_api` need the `SCHEMA` from the crate to work. Please have a look if it makes sense the way I am bringing the schema into scope here from `ockam_api::authenticator::direct`.

In those test `validate_cbor_bytes` is still used instead of `validate_cbor`. Should this be changed?

And last but not least, there are still some tests failing in `ockam_api` now where I am not entirely sure why this is the case.

Thanks for your help.

Reply:
> I am not sure where I should import the SCHEMA and create the cddl() function in ockam_api so I did in ockam_api::authenticator::direct because that's where it's used.

That should live at `schema.rs`. 

> In those test validate_cbor_bytes is still used instead of validate_cbor. Should this be changed?

Let's keep that as it is for now, just to make sure everything works as before.

> And last but not least, there are still some tests failing in ockam_api now where I am not entirely sure why this is the case.

- You have to uncomment [this line](https://github.com/build-trust/ockam/pull/3980/files#diff-1956a8258c1f4f6fe6f14fe58aace63355c293b7a09192ea689f90fdf888c087R614)
- The [schema tests form ockam_api](https://github.com/build-trust/ockam/blob/f0451afbc6b6c09e92fa9bae36da5ac50cd53d0e/implementations/rust/ockam/ockam_api/tests/schema.rs#L65) should be moved to `ockam_core` because that's were we now have the types that are used in those tests

Reply:
[schema test ](https://github.com/build-trust/ockam/blob/f0451afbc6b6c09e92fa9bae36da5ac50cd53d0e/implementations/rust/ockam/ockam_api/tests/schema.rs#L65) from `ockam_api` has been moved to `ockam_core::api` at the bottom as a test module. Not sure if that makes sense, as it's a bit cluttered now. But there is no dedicated test directory in `ockam_core`.

> That should live at schema.rs.

Not sure if you meant `ockam_api/tests/schema.rs `but I figured that does not make sense so I created a new file in `ockam_api`.

Should the `merged_cddl/ cddl` code in `ockam_core::api` also be moved to a `schema.rs` file within the crate to be consistent with `ockam_api`. Or just keep it as is?

Reply:
> [schema test ](https://github.com/build-trust/ockam/blob/f0451afbc6b6c09e92fa9bae36da5ac50cd53d0e/implementations/rust/ockam/ockam_api/tests/schema.rs#L65)from ockam_api has been moved to ockam_core::api at the bottom as a test module. Not sure if that makes sense, as it's a bit cluttered now. But there is no dedicated test directory in ockam_core.

Agree, the current code organization of the `ockam_core/api.rs` looks weird. I wouldn't worry about that now, let's land this PR first and see how it all feels. We can do that refactor later. 

> Not sure if you meant ockam_api/tests/schema.rs but I figured that does not make sense so I created a new file in ockam_api.

Sorry, I didn't have that ambiguity in mind when I wrote that. And yes, I meant creating that new file at `ockam_api/schema.rs`.

> Should the merged_cddl/ cddl code in ockam_core::api also be moved to a schema.rs file within the crate to be consistent with ockam_api. Or just keep it as is? Same with ockam_node

Not sure how to approach this right now, but I can think of a couple of different ways to organize this. We can open a different issue for this refactor and discuss the details there if that makes sense to you.